plugins/outputs/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,236 @@
+
+	// Imports the Stackdriver Monitoring client package.
+	monitoring "cloud.google.com/go/monitoring/apiv3"
+	googlepb "github.com/golang/protobuf/ptypes/timestamp"
+	metricpb "google.golang.org/genproto/googleapis/api/metric"
+	monitoredrespb "google.golang.org/genproto/googleapis/api/monitoredres"
+	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
+)
+
+// GCPStackdriver is the Google Stackdriver config info.
+type GCPStackdriver struct {
+	Project   string
+	Namespace string
+
+	client *monitoring.MetricClient
+}
+
+const (
+	// StartTime for cumulative metrics.
+	StartTime = int64(1)
+	// MaxInt is the max int64 value.
+	MaxInt = int(^uint(0) >> 1)
+)
+
+var sampleConfig = `
+  # GCP Project
+  project = "erudite-bloom-151019"
+
+  # The namespace for the metric descriptor
+  namespace = "telegraf"
+`
+
+// Connect initiates the primary connection to the GCP project.
+func (s *GCPStackdriver) Connect() error {
+	if s.Project == "" {
+		return fmt.Errorf("Project is a required field for stackdriver output")
+	}
+
+	if s.Namespace == "" {
+		return fmt.Errorf("Namespace is a required field for stackdriver output")
+	}
+
+	if s.client == nil {
+		ctx := context.Background()
+
+		// Creates a client
+		client, err := monitoring.NewMetricClient(ctx)
+		if err != nil {
+			return err
+		}
+
+		s.client = client
+	}
+
+	return nil
+}
+
+// Write the metrics to Google Cloud Stackdriver.
+func (s *GCPStackdriver) Write(metrics []telegraf.Metric) error {
+	ctx := context.Background()
+
+	for _, m := range metrics {
+		for k, v := range m.Fields() {
+			value, err := getStackdriverTypedValue(v)
+			if err != nil {
+				log.Printf("E! Error writing to output [stackdriver]: %s", err)
+				continue
+			}
+
+			metricKind, err := getStackdriverMetricKind(telegraf.Histogram)
+			if err != nil {
+				log.Printf("E! Error writing to output [stackdriver]: %s", err)
+				continue
+			}
+
+			timeInterval, err := getStackdriverTimeInterval(metricKind, StartTime, m.Time().Unix())
+			if err != nil {
+				log.Printf("E! Error writing to output [stackdriver]: %s", err)
+				continue
+			}
+
+			// Prepare an individual data point.
+			dataPoint := &monitoringpb.Point{
+				Interval: timeInterval,
+				Value:    value,
+			}
+
+			// Prepare time series.
+			timeSeries := &monitoringpb.CreateTimeSeriesRequest{
+				Name: monitoring.MetricProjectPath(s.Project),
+				TimeSeries: []*monitoringpb.TimeSeries{
+					{
+						Metric: &metricpb.Metric{
+							Type:   path.Join("custom.googleapis.com", s.Namespace, m.Name(), k),
+							Labels: m.Tags(),
+						},
+						MetricKind: metricKind,
+						Resource: &monitoredrespb.MonitoredResource{
+							Type: "global",
+							Labels: map[string]string{
+								"project_id": s.Project,
+							},
+						},
+						Points: []*monitoringpb.Point{
+							dataPoint,
+						},
+					},
+				}}
+
+			// Create the time series in Stackdriver.
+			err = s.client.CreateTimeSeries(ctx, timeSeries)
```

Comment:
Curious how the performance is here, if this is indeed one rpc call per field it could be quite a few calls.  I also worry about the expensive of the code above at a per field basis.  What has your testing shown?

Reply:
I agree, build out a `[]*monitoringpb.TimeSeries` and submit it at the end of the loop.